server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java

```diff
@@ -239,33 +385,136 @@ public void onFailure(Exception e) {
                 for (BulkItemRequest request : bulkShardRequest.items()) {
                     final String indexName = request.index();
                     DocWriteRequest<?> docWriteRequest = request.request();
-                    BulkItemResponse.Failure failure = new BulkItemResponse.Failure(indexName, docWriteRequest.id(), e);
-                    responses.set(request.id(), BulkItemResponse.failure(request.id(), docWriteRequest.opType(), failure));
+
+                    String failureStoreReference = getRedirectTarget(docWriteRequest, getClusterState().metadata());
+                    if (failureStoreReference != null) {
+                        addDocumentToRedirectRequests(request, e, failureStoreReference);
+                    }
+
+                    // Check for any previously recorded responses in the buffer since this might be a failure during a redirect
+                    BulkItemResponse response = responses.get(request.id());
+                    if (response == null) {
+                        BulkItemResponse.Failure failure = new BulkItemResponse.Failure(indexName, docWriteRequest.id(), e);
+                        response = BulkItemResponse.failure(request.id(), docWriteRequest.opType(), failure);
+                    } else {
+                        // Response already recorded. We should only be here if the existing response is a failure and
+                        // we are encountering a new failure while redirecting.
+                        assert response.isFailed() : "Attempting to overwrite successful bulk item result with a failure";
+                        response.getFailure().getCause().addSuppressed(e);
+                    }
+                    // Always replace the item in the responses for thread visibility of any mutations
+                    responses.set(request.id(), response);
```

Comment:
Is there a reason you don't use `BulkOperation#addFailure` here? It seems to match that method, except for the last line of the method.